### DISCUSSION TITLE
Include Cargo.lock into sdist

### DISCUSSION THREAD

**Discussion by hombit** [PyO3/maturin#748] [Q&A] [answered]
I'd like to make the package build more reproducible, because I've already seen how my package doesn't build because of some update of a transitive dependency. I've found that I can add
```toml
[tool.maturin]
cargo-extra-args = "--frozen --locked"
```
to prevent Cargo.lock update. But `Cargo.lock` itself is missed in a sdist, is there any way to include there for a mixed pyo3 project w/o any binaries?

**Comment by messense** [comment]
Were you putting `Cargo.lock` in `.gitignore`? I think it's automatically included in sdist if it's not ignored. You can run `cargo package --list --allow-dirty` to see the list of files to be included in sdist.

**Reply by messense** [reply]
BTW consider committing the `Cargo.lock` file to git.

**Reply by hombit** [reply]
Thanks! It is committed but it is still not presented neither in sdist nor in the `cargo package --list --allow-dirty` output. `.gitignore` doesn't list it. Could it be the issue that my package is in non-root folder of a repo?

JIC, here is the git branch I use for these experiments:
https://github.com/light-curve/light-curve-python/tree/cargo-lock
And GH Action showing that it cannot see `Cargo.lock`:
https://github.com/light-curve/light-curve-python/runs/4563451387?check_suite_focus=true

**Reply by messense** [reply]
Well that explains it, [cargo only puts `Cargo.lock` in workspace root](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html#creating-a-workspace) so it's not in  `cargo package --list --allow-dirty` of a subpackage.

I'm not sure what maturin can do right now, will need to investigate.

**Reply by hombit** [reply]
There is no workspace in this project and repo and I haven't moved Cargo.lock there, it was created by cargo and located just near Cargo.toml

**Reply by messense** [reply]
Oh I missed that. Now why `cargo package --list --allow-dirty` doesn't list `Cargo.lock`? Is it related to `Cargo.toml` being a symlink?

**Reply by messense** [reply]
```bash
light-curve-python/light-curve-python on  cargo-lock is 📦 v0.5.4-beta.0 via 🐍 v3.9.2 via 🦀 v1.57.0
❯ cargo b
error: failed to parse manifest at `/Users/messense/Projects/light-curve-python/light-curve-python/Cargo.toml`

Caused by:
  can't find library `light_curve`, rename file to `src/lib.rs` or specify lib.path
```

That branch doesn't even build.

**Reply by messense** [reply]
OK, I looked at Cargo's source code and it looks it `cargo package` ignore `Cargo.lock` unless there is an example or binary target.

1. https://github.com/rust-lang/cargo/blob/3787309df66b32995f91bbf4d251e17e90fca4af/src/cargo/ops/cargo_package.rs#L230
2. https://github.com/rust-lang/cargo/blob/3787309df66b32995f91bbf4d251e17e90fca4af/src/cargo/ops/cargo_package.rs#L245-L251
3. https://github.com/rust-lang/cargo/blob/2e2a16e983f597da62bc132eb191bc3276d4b1bb/src/cargo/core/package.rs#L206-L209

**Reply by hombit** [reply]
Thank you so much!

**Comment by messense** [comment, marked as answer]
I think you can add `Cargo.lock` to it by adding the following content in your `pyproject.toml`

```toml
[tool.maturin]
sdist-include = ["Cargo.lock"]
```

But I do think maturin should do it automatically if user specified `--locked` or `--frozen` in `cargo-extra-args`.

**Reply by messense** [reply]
Implemented in https://github.com/PyO3/maturin/pull/749

**Reply by hombit** [reply]
I've checked it works like a charm!
